src/main/java/appeng/fluids/helper/DualityFluidInterface.java

```diff
@@ -0,0 +1,632 @@
+	}
+
+	private void updatePlan( final int slot )
+	{
+		final IAEFluidStack req = this.configStacks[slot];
+		final FluidStack stored = this.tanks[slot].drain( TANK_CAPACITY, false );
+
+		if( req == null && ( stored != null && stored.amount > 0 ) )
+		{
+			final IAEFluidStack work = AEApi.instance().storage().getStorageChannel( IFluidStorageChannel.class ).createStack( stored );
+			this.requireWork[slot] = work.setStackSize( -work.getStackSize() );
+			return;
+		}
+		else if( req != null )
+		{
+			if( stored == null || stored.amount == 0 ) // need to add stuff!
+			{
+				this.requireWork[slot] = req.copy();
+				return;
+			}
+			else if( req.getFluid().equals( stored.getFluid() ) ) // same type ( qty different? )!
+			{
+				if( stored.amount < TANK_CAPACITY )
+				{
+					this.requireWork[slot] = req.copy();
+					this.requireWork[slot].setStackSize( TANK_CAPACITY - stored.amount );
+					return;
+				}
+			}
+			else
+			// Stored != null; dispose!
+			{
+				final IAEFluidStack work = AEApi.instance().storage().getStorageChannel( IFluidStorageChannel.class ).createStack( stored );
+				this.requireWork[slot] = work.setStackSize( -work.getStackSize() );
+				return;
+			}
+		}
+
+		this.requireWork[slot] = null;
+	}
+
+	private boolean usePlan( final int slot )
+	{
+		IFluidHandler tank = this.tanks[slot];
+		IAEFluidStack work = this.requireWork[slot];
+		this.isWorking = slot;
+
+		boolean changed = false;
+		try
+		{
+			final IMEInventory<IAEFluidStack> dest = this.gridProxy.getStorage()
+					.getInventory( AEApi.instance().storage().getStorageChannel( IFluidStorageChannel.class ) );
+			final IEnergySource src = this.gridProxy.getEnergy();
+
+			if( work.getStackSize() > 0 )
+			{
+				// make sure strange things didn't happen...
+				if( tank.fill( work.getFluidStack(), false ) != work.getStackSize() )
+				{
+					changed = true;
+				}
+				else
+				{
+					final IAEFluidStack acquired = Platform.poweredExtraction( src, dest, work, this.interfaceRequestSource );
+					if( acquired != null )
+					{
+						changed = true;
+						final int filled = tank.fill( acquired.getFluidStack(), true );
+						if( filled != acquired.getStackSize() )
+						{
+							throw new IllegalStateException( "bad attempt at managing tanks. ( fill )" );
+						}
+					}
+				}
+			}
+			else if( work.getStackSize() < 0 )
+			{
+				IAEFluidStack toStore = work.copy();
+				toStore.setStackSize( -toStore.getStackSize() );
+
+				// make sure strange things didn't happen...
+				final FluidStack canExtract = tank.drain( toStore.getFluidStack(), false );
+				if( canExtract == null || canExtract.amount != toStore.getStackSize() )
+				{
+					changed = true;
+				}
+				else
+				{
+					IAEFluidStack notStored = Platform.poweredInsert( src, dest, toStore, this.interfaceRequestSource );
+					toStore.setStackSize( toStore.getStackSize() - (notStored == null ? 0 : notStored.getStackSize()));
+
+					if( toStore.getStackSize() > 0 )
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS134)